engine/verification/assigner/engine.go

```diff
@@ -0,0 +1,255 @@
+)
+
+// The Assigner engine reads the receipts from each finalized block.
+// For each receipt, it reads its result and find the chunks the assigned
+// to me to verify, and then save it to the chunks job queue for the
+// fetcher engine to process.
+type Engine struct {
+	unit               *engine.Unit
+	log                zerolog.Logger
+	metrics            module.VerificationMetrics
+	tracer             module.Tracer
+	me                 module.Local
+	state              protocol.State
+	assigner           module.ChunkAssigner  // to determine chunks this node should verify.
+	chunksQueue        storage.ChunksQueue   // to store chunks to be verified.
+	newChunkListener   module.NewJobListener // to notify chunk queue consumer about a new chunk.
+	processingNotifier ProcessingNotifier    // to report a block has been processed.
+	indexer            storage.Indexer       // to index receipts of a block based on their executor identifier.
+}
+
+func New(
+	log zerolog.Logger,
+	metrics module.VerificationMetrics,
+	tracer module.Tracer,
+	me module.Local,
+	state protocol.State,
+	assigner module.ChunkAssigner,
+	chunksQueue storage.ChunksQueue,
+	newChunkListener module.NewJobListener,
+	indexer storage.Indexer,
+) *Engine {
+	return &Engine{
+		unit:             engine.NewUnit(),
+		log:              log.With().Str("engine", "assigner").Logger(),
+		metrics:          metrics,
+		tracer:           tracer,
+		me:               me,
+		state:            state,
+		assigner:         assigner,
+		chunksQueue:      chunksQueue,
+		newChunkListener: newChunkListener,
+		indexer:          indexer,
+	}
+}
+
+// handleExecutionReceipt receives a receipt that appears in a finalized container block. In case this verification node
+// is staked at the reference block of this execution receipt's result, chunk assignment is done on the execution result, and
+// the assigned chunks' locators are pushed to the chunks queue, which are made available for the chunk queue consumer (i.e., the
+// fetcher engine).
+func (e *Engine) handleExecutionReceipt(receipt *flow.ExecutionReceipt, containerBlockID flow.Identifier) {
+	receiptID := receipt.ID()
+	resultID := receipt.ExecutionResult.ID()
+	referenceBlockID := receipt.ExecutionResult.BlockID
+
+	log := log.With().
+		Hex("receipt_id", logging.ID(receiptID)).
+		Hex("result_id", logging.ID(resultID)).
+		Hex("reference_block_id", logging.ID(referenceBlockID)).
+		Hex("container_block_id", logging.ID(containerBlockID)).
+		Logger()
+
+	// verification node should be staked at the reference block id.
+	ok, err := e.stakedAtBlockID(receipt.ExecutionResult.BlockID)
+	if err != nil {
+		log.Error().Err(err).Msg("could not verify stake of verification node for result")
+		return
+	}
+	if !ok {
+		log.Debug().Msg("node is not staked at reference block id, receipt is discarded")
+		return
+	}
+
+	// chunk assignment
+	chunkList, err := e.chunkAssignments(context.Background(), &receipt.ExecutionResult)
```

Comment:
For this PR, we let `chunkAssignments` reside on the legacy tracer of `match` engine. Wiring the tracer and metrics modules are the subject of the next PR.

Reply:
**Note**: `chunkAssigner` has an internal mempool `cache`, i.e., it computes the assignment only once for each result, and serves the subsequent calls on the same _input_ with the same _assignment_. Hence, even with duplicate results, running `chunkAssigner` does not apply a noticeable computational overhead.